libcfnet/protocol.h

```diff
@@ -64,4 +64,55 @@
  */
 Seq *ProtocolOpenDir(AgentConnection *conn, const char *path);
 
+/**
+ * Receives a file from a remote host.
+ *
+ * The server will use "/var/cfengine" as working directory if absolute path
+ * is not specified. The client will send a request that looks like this:
+ *      `GET <buf_size> <remote_path>`
+ *
+ * `buf_size` is the local buffer size: how much of the file to receive in
+ * each transaction. It should be aligned to block size (which is usually
+ * 4096), but currently the protocol only allows _exactly_ 4095 bytes per
+ * transaction.
+ *
+ * The function shall fail if connection is not established, or if the server
+ * gives a bad response (denoted by a message preceded by "BAD").
+ *
+ * @param [in] conn         The connection to use
+ * @param [in] remote_path  Path on remote host
+ * @param [in] local_path   Path of received file
+ * @param [in] file_size    Size of file to get
+ * @param [in] perms        Permissions of local file
+ * @return True if file was successfully transferred, false otherwise
+ *
+ * Example (for printing each directory entry):
+ * @code
+ *     AgentConnection *conn = ServerConnection("127.0.0.1", "666", ...);
+ *     bool got_file = ProtocolGet(conn, "masterfiles/update.cf",
+ *                                 "update.cf", CF_MSGSIZE, 0644);
+ *     if (got_file)
+ *     {
+ *        struct stat sb;
+ *        stat("update.cf", &sb);
+ *        printf("This file is %ld big!\n", sb.st_size);
+ *     }
+ * @endcode
+ *
+ * In the protocol, this will look like this on the server side:
+ *     Received: GET masterfiles/update.cf
+ *     Translated to: GET /var/cfengine/masterfiles/update.cf
+ */
+bool ProtocolGet(AgentConnection *conn, const char *remote_path,
+                 const char *local_path, const uint32_t file_size, int perms);
```

Comment:
No big deal, but I don't think the `uint32_t` should be declared `const` in the declaration. IMO, it only makes sense to include it in the function definition (implementation).

Reply:
Considering the documentation is in the header file, I think it provides useful information that the function will not use another file size than the one provided by the caller.

Reply:
No. `const` in the declaration declares that you won't change the parameter. For `const char *`, this is useful because you can reasonably assume that a function with this signature won't change your string (the data the pointer points to). It in no way guarantees that the parameter `file_size` will be used in the protocol command, nothing stops the function from falling back to something else, for example:

```C
snprintf(buf, sizeof(buf), "GET %d /my/file", file_size ? file_size : 32);
```

`const int` or similar in the declaration is confusing, it does nothing, and implies that the function would be able to alter your int, if the `const` was not there. 

In the definiton (implementation), it does have a function:

```C
bool ProtocolGet(const char *const path, const int file_size)
{
  [... big function with lots of logic ...]
  // Since path and file_size are const, we know that the spaghetti above 
  // didn't change them. We've also prevented someone from altering them
  // perhaps accidentally.
  Log(LOG_LEVEL_INFO, "Got file '%s' of size %d", path, file_size);
  return true;
}
```

Reply:
> It in no way guarantees that the parameter file_size will be used in the protocol command, nothing stops the function from falling back to something else

It's hard to not agree with this point, but I would still argue that it still adds understanding. I like to match parameters in declarations and definitions.

> In the definiton (implementation), it does have a function:

Yes, I am aware of that. But regarding the point that it does not do anything in the declaration: you might as well say that adding parameter names in the declaration is pointless because they don't have a function there.

Reply:
@karlhto parameter names clearly have a purpose, they explain what the parameter means (a human readable name). I don't think anyone would argue this. However your interpretation of what `const` means is arbitrary, it's not based on how the language works, or what the const keyword usually means.

> It's hard to not agree with this point, but I would still argue that it still adds understanding. 

And I think it adds confusion. 

> I like to match parameters in declarations and definitions.

We've discussed this in another PR before, with craig I believe. Yes, you can sync them, that's kind of a different point than what you were suggesting above. Above, you stated that `const` adds understanding. If you really believe that, you could start adding more `const` to declarations where it's not even in the definitons.

That's fine for this case. When you get a lot of consts in the function declaration, I think it makes it a lot more easy to read if you drop the unnecessary ones. There are even some quite rare cases where I would say it's appropriate for the function definition to use different parameter names from the declaration.

Reply:
+1 to @olehermanse from me

Reply:
> it still provides a hint that another value won't be used (in my opinion).

@karlhto It doesn't. This is where we disagree. That's just not what `const` means. Within the function body, it means that the local variable `file_size` is constant, cannot be assigned to. Again, in the declaration it is just confusing. (If I saw that I would wonder why it was there, not assume that it was what you are hinting at).

> It's not a big deal and we probably shouldn't spend this much energy on this.

Yes. I wouldn't block your PR on this, but since this is the second time it comes up during PR review, we should probably "settle" it and add a short guideline to our style guide.

Reply:
> we should probably "settle" it and add a short guideline to our style guide.

@olehermanse Sounds good. I am fine with settling on your style since @vpodzime seems to agree.